core/src/test/scala/unit/kafka/controller/ReassignmentsManagerTest.scala

```diff
@@ -0,0 +1,228 @@
+    val results = partitionReassignmentManager.triggerApiReassignment(Map(tp -> Some(Seq(1,2,3))))
+    assertTrue(results(tp).is(Errors.UNKNOWN_TOPIC_OR_PARTITION))
+  }
+
+  /**
+   * Phase A of a partition reassignment denotes the initial trigger of a reassignment.
+   *
+   * A1. Bump the leader epoch for the partition and send LeaderAndIsr updates to RS.
+   * A2. Start new replicas AR by moving replicas in AR to NewReplica state.
+   */
+  @Test
+  def testPhaseAOfPartitionReassignment(): Unit = {
+    /*
+     * Existing assignment is [1,2,3]
+     * We issue a reassignment to [3, 4, 5]
+     */
+    val expectedAddingReplicas = Seq(4, 5)
+    val expectedFullReplicaSet = Seq(3, 4, 5, 1, 2)
+    val initialAssignment = ReplicaAssignment(Seq(1, 2, 3), Seq(), Seq())
+    val initialLeaderAndIsr = new LeaderAndIsr(1, 1, List(1, 2, 3), zkEpoch)
+    controllerContext.partitionAssignments.put(topic, mutable.Map(tp.partition() -> initialAssignment))
+    controllerContext.updatePartitionFullReplicaAssignment(tp, initialAssignment)
+    controllerContext.partitionsBeingReassigned.add(tp)
+    setLiveBrokers(Seq(1,2,3,4,5))
+    mockAreReplicasInIsr(tp, List(1, 2, 3), initialLeaderAndIsr)
+    val expectedNewAssignment = ReplicaAssignment.fromOldAndNewReplicas(Seq(1, 2, 3), Seq(3, 4, 5))
+    assertEquals(expectedFullReplicaSet, expectedNewAssignment.replicas)
+    // U1. Should update ZK
+    doReturn(mockSetDataResponseOK, Nil: _*).when(mockZkClient).setTopicAssignmentRaw(topic, mutable.Map(tp -> expectedNewAssignment), zkEpoch)
+    // U2. Should update memory
+    // A1. Should update partition leader epoch in ZK
+    val expectedLeaderAndIsr = initialLeaderAndIsr.newEpochAndZkVersion
+    doReturn(UpdateLeaderAndIsrResult(Map(tp -> Right(expectedLeaderAndIsr)), Seq()), Nil: _*)
+      .when(mockZkClient).updateLeaderAndIsr(Map(tp -> expectedLeaderAndIsr), controllerEpoch, zkEpoch)
+
+    // act
+    val results = partitionReassignmentManager.triggerApiReassignment(Map(tp -> Some(Seq(3, 4, 5))))
+    assertTrue(s"reassignment failed - $results", results(tp).isSuccess)
+
+     // U2. Should have updated memory
+    assertEquals(expectedNewAssignment, controllerContext.partitionFullReplicaAssignment(tp))
+    // A1. Should send a LeaderAndIsr request to every replica in ORS + TRS (with the new RS, AR and RR).
+    verify(mockBrokerRequestBatch).addLeaderAndIsrRequestForBrokers(
+      expectedFullReplicaSet, tp,
+      LeaderIsrAndControllerEpoch(expectedLeaderAndIsr, controllerEpoch), expectedNewAssignment, isNew = false
+    )
+    // A2. replicas in AR -> NewReplica
+    expectedAddingReplicas.foreach { newReplica =>
+      verify(mockReplicaStateMachine).handleStateChanges(
+        Seq(PartitionAndReplica(tp, newReplica)), NewReplica
+      )
+    }
+    verify(mockBrokerRequestBatch).sendRequestsToBrokers(controllerEpoch)
+  }
+
+  /**
+   * Phase B of a partition reassignment is the part where all the new replicas are in ISR
```

Comment:
Is this duplication necessary? It seems likely to diverge over time.

Reply:
Theoretically, if we change the code these tests would fail and this gets updated but I can see how it's likely to get missed. Let's keep the comments in between the test code though